editor/src/components/inspector/fill-hug-fixed-control.tsx

```diff
@@ -817,12 +809,80 @@ function setGroupChildConstraint(
 
   const actions: EditorAction[] = mapDropNulls((path) => {
     const constraints = getSafeGroupChildConstraintsArray(allElementProps, path)
-    const newProps = makeNewProps(constraints, mode)
-    const uniqueNewProps = uniqBy(newProps, (a, b) => a === b)
-    return uniqueNewProps.length === 0
+    const newProps = makeUpdatedConstraintsPropArray(constraints, mode, dimension)
+    return newProps.length === 0
       ? unsetProperty(path, prop)
-      : setProp_UNSAFE(path, prop, jsExpressionValue(uniqueNewProps, emptyComments))
+      : setProp_UNSAFE(path, prop, jsExpressionValue(newProps, emptyComments))
   }, selectedViews)
 
   dispatch(actions)
 }
+
+export function makeUpdatedConstraintsPropArray(
+  constraints: any[],
+  mode: ConstraintsMode,
+  dimension: LayoutPinnedProp,
+): any[] {
+  const newProps = new Set(constraints)
+  switch (mode) {
+    case 'add':
+      newProps.add(dimension)
+      break
+    case 'remove':
+      newProps.delete(dimension)
+      break
+    default:
+      assertNever(mode)
+  }
+
+  if (mode === 'add') {
+    switch (dimension) {
+      case 'width':
+        maybeRemoveOldestComplement(newProps, constraints, ['left', 'right'])
+        break
+      case 'height':
+        maybeRemoveOldestComplement(newProps, constraints, ['top', 'bottom'])
+        break
+      case 'top':
+        maybeRemoveComplementaryDimension(newProps, 'bottom', 'height')
+        break
+      case 'left':
+        maybeRemoveComplementaryDimension(newProps, 'right', 'width')
+        break
+      case 'bottom':
+        maybeRemoveComplementaryDimension(newProps, 'top', 'height')
+        break
+      case 'right':
+        maybeRemoveComplementaryDimension(newProps, 'left', 'width')
+        break
+      default:
+        assertNever(dimension)
+    }
+  }
+
+  return Array.from(newProps)
+}
+
+function maybeRemoveOldestComplement(
+  set: Set<any>,
```

Comment:
since we set the `'data-constraints'` prop via the tool, I think it'd be worth it to validate its value `getSafeGroupChildConstraintsArray` so that we only take into account the valid values, and ignore whatever else that might have made its way into the prop (and get rid the `any`s stemming from `getSafeGroupChildConstraintsArray`

Reply:
makes sense, done in [`6f34c44` (#4282)](https://github.com/concrete-utopia/utopia/pull/4282/commits/6f34c44590fdfe7ef31f57915b07e2e24555672b)